lib/ecto/adapters/clickhouse/connection.ex

```diff
@@ -914,9 +922,35 @@ defmodule Ecto.Adapters.ClickHouse.Connection do
       message: "unknown or ambiguous (for ClickHouse) Ecto type #{inspect(type)}"
   end
 
-  defp param_type_at(params, ix) do
-    value = Enum.at(params, ix)
-    param_type(value)
+  defp interpolate_param(nil), do: "NULL"
+  defp interpolate_param(true), do: "true"
+  defp interpolate_param(false), do: "false"
+  defp interpolate_param(s) when is_binary(s), do: [?', escape_string(s), ?']
+  defp interpolate_param(i) when is_integer(i), do: Integer.to_string(i)
+  # TODO ensure not scientific notation, or ensure ClickHouse can read it
+  defp interpolate_param(f) when is_float(f), do: Float.to_string(f)
+  defp interpolate_param(%NaiveDateTime{} = naive), do: [?', NaiveDateTime.to_string(naive), ?']
+  # TODO DateTime64
+  defp interpolate_param(%DateTime{} = dt), do: Integer.to_string(DateTime.to_unix(dt, :second))
+  defp interpolate_param(%Date{} = date), do: [?', Date.to_string(date), ?']
+  defp interpolate_param(%Decimal{} = dec), do: Decimal.to_string(dec, :normal)
+
+  defp interpolate_param(a) when is_list(a) do
+    [?[, Enum.map_intersperse(a, ?,, &interpolate_param/1), ?]]
+  end
+
+  defp interpolate_param(%s{}) do
+    raise ArgumentError, "struct #{inspect(s)} is not supported in params"
+  end
+
+  defp interpolate_param(m) when is_map(m) do
+    [
+      ?{,
+      Enum.map_intersperse(m, ?,, fn {k, v} ->
+        [interpolate_param(k), ?:, interpolate_param(v)]
+      end),
+      ?}
+    ]
```

Comment:
These clauses are somewhat similar to https://github.com/plausible/ch/blob/2ed12be5c7254531b1decefd95e2f5da01a6a17f/lib/ch/query.ex#L208-L294, maybe we'd be able to reuse it. For now I'm just duplicating.